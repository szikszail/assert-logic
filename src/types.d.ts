
export type AnyValue = void | boolean | string | object | number | null;
export type AsyncAnyValue = Promise<AnyValue>;
export type AssertionFunction = () => AnyValue | AsyncAnyValue;
export type AssertionValue = AssertionFunction | AnyValue | AsyncAnyValue;
