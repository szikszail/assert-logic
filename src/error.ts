import {Assertion} from "./assertion/assertion";
import {stringify} from "yaml";
import {AnyValue, EvaluationResult} from "./types";
import {valueToString} from "./assertion/pass";

type ErrorRepr = Record<string, string | (ErrorRepr | string)[]>;

function getErrorRepresentation(operator: string, expectation: string, assertion: Assertion, results?: EvaluationResult[]): ErrorRepr {
  const repr: ErrorRepr = {
    [`AssertionError (${operator})`]: expectation,
  };
  if (results?.length > 0) {
    repr.Results = results.map((result) => {
      if (result instanceof AssertionError) {
        return getErrorRepresentation(result.operator, result.expectation, result.assertion, result.results);
      }
      if (result instanceof Error) {
        return {
          // eslint-disable-next-line no-control-regex
          Error: result.message.replace(/\u001b\[\d+m/g, ''),
        };
      }
      return "Pass";
    });
  }
  repr.Expression = assertion.toString();
  return repr;
}

export class AssertionError extends Error {
  constructor(
    public readonly operator: string,
    public readonly expectation: string,
    public readonly assertion: Assertion,
    public readonly results?: EvaluationResult[],
  ) {
    super(stringify(
      getErrorRepresentation(operator, expectation, assertion, results),
      {
        indent: 2,
        // defaultStringType: "PLAIN",
        // defaultKeyType: "PLAIN",
        collectionStyle: 'block',
        // lineWidth: 0,
        // blockQuote: true,
        doubleQuotedMinMultiLineLength: Infinity,
      }
    ));
    this.name = "AssertionError";
  }
}

export class FailedAssertionError extends Error {
  constructor(value: AnyValue) {
    super(`Failed expression: ${valueToString(value)}`);
    this.name = "FailedAssertionError";
  }
}