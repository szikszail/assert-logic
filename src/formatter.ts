import type {AssertionValue} from "./types";

export function toString(value: AssertionValue): string {
  if (value instanceof Function) {
    let name = value?.name;
    if (!name) {
      name = value.toString();
    }
    if (name.length > 20) {
      name = name.slice(0, 17) + "...";
    }
    return `function(${name})`;
  }
  if (value instanceof Promise) {
    return value.toString();
  }
  return `value(${value})`;
}