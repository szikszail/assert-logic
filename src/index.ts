// import debug = require("debug");

// const log = debug("assert-logic");

import type {AssertionValue} from "./types";
import {PassAssertion} from "./assertion/pass";
import {NotAssertion} from "./assertion/not";

export const pass = (value: AssertionValue) => new PassAssertion(value);
export const not = (value: AssertionValue) => new NotAssertion(value);