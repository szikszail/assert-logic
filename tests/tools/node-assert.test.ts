import {or, and} from '../../src';
import assert from 'node:assert';

describe('node:assert', () => {
  test("should work with Node:Assert assertions (positive)", () => {
    const number = 12;
    expect(() => and(
      () => assert(number, "number"),
      () => assert(number % 2 === 0),
      or(
        () => assert(number < 15),
        () => assert(number > 10),
      ),
    ).evaluate()).not.toThrow();
  });
  test("should work with Node:Assert assertions (negative)", () => {
    const number = 12;
    expect(() => and(
      () => assert(number, "number"),
      () => assert(number % 2 === 1),
      or(
        () => assert(number < 7),
        () => assert(number > 10),
      ),
    ).evaluate()).toThrowErrorMatchingSnapshot()
  });
});