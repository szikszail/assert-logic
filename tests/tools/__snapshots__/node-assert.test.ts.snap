// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`node:assert should work with Node:Assert assertions (negative) 1`] = `
"AssertionError (AND): Expected all expression to pass, but not all did.
Results:
  - Pass
  - AssertionError (PASS): Expected expression to pass.
    Results:
      - Pass
    Expression: (function)
  - Pass
Expression: |-
  AND(
    (function)
    (function)
    OR(
      (function)
      (function)
    )
  )
"
`;
